rollout/trafficrouting.go

```diff
@@ -17,152 +17,177 @@ import (
 )
 
 // NewTrafficRoutingReconciler identifies return the TrafficRouting Plugin that the rollout wants to modify
-func (c *Controller) NewTrafficRoutingReconciler(roCtx *rolloutContext) (trafficrouting.TrafficRoutingReconciler, error) {
+func (c *Controller) NewTrafficRoutingReconciler(roCtx *rolloutContext) ([]trafficrouting.TrafficRoutingReconciler, error) {
 	rollout := roCtx.rollout
+	// define an empty list of trafficReconcilers to be populated
+	// by the ones declared in the rolloutContext
+	trafficReconcilers := []trafficrouting.TrafficRoutingReconciler{}
+
 	if rollout.Spec.Strategy.Canary.TrafficRouting == nil {
 		return nil, nil
 	}
 	if rollout.Spec.Strategy.Canary.TrafficRouting.Istio != nil {
 		if c.IstioController.VirtualServiceInformer.HasSynced() {
-			return istio.NewReconciler(rollout, c.IstioController.DynamicClientSet, c.recorder, c.IstioController.VirtualServiceLister, c.IstioController.DestinationRuleLister), nil
+			trafficReconcilers = append(trafficReconcilers, istio.NewReconciler(rollout, c.IstioController.DynamicClientSet, c.recorder, c.IstioController.VirtualServiceLister, c.IstioController.DestinationRuleLister))
 		} else {
-			return istio.NewReconciler(rollout, c.IstioController.DynamicClientSet, c.recorder, nil, nil), nil
+			trafficReconcilers = append(trafficReconcilers, istio.NewReconciler(rollout, c.IstioController.DynamicClientSet, c.recorder, nil, nil))
 		}
 	}
 	if rollout.Spec.Strategy.Canary.TrafficRouting.Nginx != nil {
-		return nginx.NewReconciler(nginx.ReconcilerConfig{
+		trafficReconcilers = append(trafficReconcilers, nginx.NewReconciler(nginx.ReconcilerConfig{
 			Rollout:        rollout,
 			Client:         c.kubeclientset,
 			Recorder:       c.recorder,
 			ControllerKind: controllerKind,
 			IngressLister:  c.ingressesLister,
-		}), nil
+		}))
 	}
 	if rollout.Spec.Strategy.Canary.TrafficRouting.ALB != nil {
-		return alb.NewReconciler(alb.ReconcilerConfig{
+		alb_reconcilier, err := alb.NewReconciler(alb.ReconcilerConfig{
 			Rollout:        rollout,
 			Client:         c.kubeclientset,
 			Recorder:       c.recorder,
 			ControllerKind: controllerKind,
 			IngressLister:  c.ingressesLister,
 		})
+		trafficReconcilers = append(trafficReconcilers, alb_reconcilier)
+		if err != nil {
+			return trafficReconcilers, err
+		}
 	}
 	if rollout.Spec.Strategy.Canary.TrafficRouting.SMI != nil {
-		return smi.NewReconciler(smi.ReconcilerConfig{
+		smi_reconcilier, err := smi.NewReconciler(smi.ReconcilerConfig{
 			Rollout:        rollout,
 			Client:         c.smiclientset,
 			Recorder:       c.recorder,
 			ControllerKind: controllerKind,
 		})
+		trafficReconcilers = append(trafficReconcilers, smi_reconcilier)
```

Comment:
Same here. Maybe its better to check the error first before appending to the `trafficReconcilers`

Reply:
The errors are controlled here:
```
reconcileTrafficRouting() error {
	reconcilers, err := c.newTrafficRoutingReconciler(c)
	// a return here does ensure that all trafficReconcilers are healthy
	// and sane in syntax
	if err != nil {
		return err
	}
```
the err return is part of the original code, so even in the previous case a `[valid, nil], err` or `[nil], err` or `[valid], err` does result in the same outcome. If `err` is present, the whole reconciliation fails

Reply:
any traffic controller fails with err, it should return error and fail whole reconciliation .. in the above trafficReconcilers updates, err checks are not done for all controllers.

Reply:
change commited.